libdnf/dnf-context.cpp

```diff
@@ -3000,6 +3000,10 @@ libdnf::ConfigMain & getGlobalMainConfig()
                     }
                 }
             }
+        } catch (const libdnf::ConfigParser::CantOpenFile & ex) {
+            if (configFilePath) {
+                g_warning("Loading \"%s\": %s", cfgPath.c_str(), ex.what());
```

Comment:
Seems like this is an error that should be propagated instead of logged with a warning and ignored? In CLIs (dnf, microdnf) this should cause the command to exit with an error code (I guess my https://github.com/rpm-software-management/libdnf/pull/866 would be needed for that?).

Reply:
Only warning is logged at that place. There is not a possibility to safely log an error at this place.
But the file is opened again (for loading repositories defined in it). And there is the problem catched, an error is generated with description.
I will add a comment with that info to the code.

Example with non existent file:
```
# microdnf --config=/etc/dnf/dnf.con repolist

(microdnf:34835): libdnf-WARNING **: 07:13:01.404: Loading "/etc/dnf/dnf.con": IniParser: Can't open file
error: Failed to load /etc/dnf/dnf.con: Failed to open file ?/etc/dnf/dnf.con?: No such file or directory
# 
```
Example with missing permissions:
```
# microdnf --config=/etc/dnf/dnf.conf repolist

(microdnf:34923): libdnf-WARNING **: 07:16:17.568: Loading "/etc/dnf/dnf.conf": IniParser: Can't open file
error: Failed to load /etc/dnf/dnf.conf: Failed to open file ?/etc/dnf/dnf.conf?: Permission denied
#
```

Reply:
I went to have a look at what's actually going on in the code and I can't see a reason why you can't actually throw an exception here, at least for this particular code path being used here (there might be code paths that can't propagate errors properly, but then still you should be throwing an exception here and then deal with it accordingly in the various places...).

This is C++ code and it should be throwing exceptions in case of errors, not logging warnings and ignoring the errors.

In this case this function is called from `dnf_context_setup()`, which has a `GError **error` argument, so you can throw an exception here, catch it in `dnf_context_setup()`, convert to `GError` and propagate the error properly.

I'll also add that the new `IniParser`'s `CantOpenFile` exception doesn't contain any description of the error whatsoever, while the old config parsing code properly reports what the error with opening the config file is. That really needs to be fixed too.